### PR TITLE
[FIRRTL][OMIR] Detect and error if tracker resolution is ambiguous.

### PR DESCRIPTION
**Pull request by dtzSiFive**
Current behavior is to silently use the first we happen to discover.

Reachable by putting OMIRTracker on an aggregate, which LowerTypes will clone out to the lowered elements (?).

Presently, in practice, we tend to resolve these trackers to the first element of the aggregate.

### CODE REVIEW

File: lib/Dialect/FIRRTL/Transforms/EmitOMIR.cpp [hunk 1]
```diff
@@ -640,7 +640,16 @@ void EmitOMIRPass::runOnOperation() {
       }
       if (sramIDs.erase(tracker.id))
         makeTrackerAbsolute(tracker);
-      trackers.insert({tracker.id, tracker});
+      if (auto [it, inserted] = trackers.try_emplace(tracker.id, tracker);
+          !inserted) {
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/Transforms/EmitOMIR.cpp:644]: It took me a second to parse this syntax. Nice and terse.

File: test/Dialect/FIRRTL/emit-omir-errors.mlir [hunk 2]
```diff
@@ -69,3 +69,20 @@ firrtl.circuit "Foo" attributes {annotations = [{
...
+  nodes = [{info = #loc, id = "OMID:0", fields = {
+    a = {info = #loc, index = 1, value = {omir.tracker, id = 0, path = "~Foo|Foo>n", type = "OMMemberReferenceTarget"}}
+  }}]
+}]} {
+  firrtl.module @Foo(in %x_a: !firrtl.uint<1>, in %x_b: !firrtl.uint<1>) {
+    // expected-note @below {{tracker with same ID already found here}}
+    %n_a = firrtl.node %x_a {annotations = [{class = "freechips.rocketchip.objectmodel.OMIRTracker", id = 0 : i64, type = "OMMemberReferenceTarget"}]} : !firrtl.uint<1>
+    // expected-error @below {{freechips.rocketchip.objectmodel.OMIRTracker annotation with same ID already found, must resolve to single target}}
+    %n_b = firrtl.node %x_b {annotations = [{class = "freechips.rocketchip.objectmodel.OMIRTracker", id = 0 : i64, type = "OMMemberReferenceTarget"}]} : !firrtl.uint<1>
+}
```
Comment (seldridge) [inline test/Dialect/FIRRTL/emit-omir-errors.mlir:87]: Nit: using an uninitialized wire would avoid having to have ports.

**Review by seldridge — APPROVED** [review]
LGTM

Thanks for tightening up the behavior. Any spot checking on internal designs to see if this is unintentionally being used?

File: test/Dialect/FIRRTL/emit-omir-errors.mlir (hunk 2 above)
Comment (dtzSiFive) [reply test/Dialect/FIRRTL/emit-omir-errors.mlir]: Updated, thanks!

**Comment by dtzSiFive** [comment]
> Thanks for tightening up the behavior. Any spot checking on internal designs to see if this is unintentionally being used?

Ran few designs and seems unused (no errors/breakages anyway, in theory that could mean only one element of the aggregate reaches this point but that seems unlikely).  There's a related pattern of dontTouch + OMIR tracking which individually marks aggregate elements as some sort of workaround, but not 100% if that's working around this aspect or something else.  Hopefully rejecting will help avoid unexpected output should any new uses be introduced.

**Comment by dtzSiFive** [comment]
Thanks for the review(s)!

**Pull request merged**